src/vmm/src/acpi/mod.rs

```diff
@@ -0,0 +1,219 @@
+            OEM_REVISION,
+            self.get_apic_addr(),
+            self.setup_interrupt_controllers(nr_vcpus),
+        );
+        self.write_acpi_table(&mut madt)
+    }
+
+    fn build_xsdt(&mut self, fadt_addr: u64, madt_addr: u64) -> Result<u64, AcpiError> {
+        debug!("acpi: Building XSDT table");
+        let mut xsdt = Xsdt::new(
+            OEM_ID,
+            *b"FCMVXSDT",
+            OEM_REVISION,
+            vec![fadt_addr, madt_addr],
+        );
+        self.write_acpi_table(&mut xsdt)
+    }
+
+    fn build_rsdp(&mut self, xsdt_addr: u64) -> Result<(), AcpiError> {
+        debug!("acpi: Building RSDP");
+        let mut rsdp = Rsdp::new(OEM_ID, xsdt_addr);
+        rsdp.write_to_guest(&self.guest_memory, self.rsdp_addr())
+            .inspect_err(|err| debug!("acpi: Could not write RSDP in guest memory: {err}"))?;
+
+        debug!(
+            "acpi: Wrote RSDP ({} bytes) at address: {:#010x}",
+            rsdp.len(),
+            self.rsdp_addr().0
+        );
+        Ok(())
+    }
+
+    pub(crate) fn create_acpi_tables(&mut self, vcpus: &[Vcpu]) -> Result<(), AcpiError> {
+        let dsdt_addr = self.build_dsdt()?;
+        let fadt_addr = self.build_fadt(dsdt_addr)?;
+        let madt_addr = self.build_madt(vcpus.len().try_into().unwrap())?;
+        let xsdt_addr = self.build_xsdt(fadt_addr, madt_addr)?;
+        self.build_rsdp(xsdt_addr)
+    }
+}
+
+#[cfg(test)]
+pub mod tests {
+    use acpi_tables::Sdt;
+    use vm_memory::Bytes;
+
+    use crate::acpi::AcpiError;
+    use crate::arch::x86_64::layout::{ACPI_MEM_SIZE, ACPI_MEM_START};
+    use crate::builder::tests::default_vmm;
+    use crate::utilities::test_utils::arch_mem;
+
+    struct MockSdt(Vec<u8>);
+
+    impl Sdt for MockSdt {
+        fn len(&self) -> usize {
+            self.0.len()
+        }
+
+        fn write_to_guest<M: vm_memory::GuestMemory>(
+            &mut self,
+            mem: &M,
+            address: vm_memory::GuestAddress,
+        ) -> acpi_tables::Result<()> {
+            mem.write_slice(&self.0, address)?;
+            Ok(())
+        }
+    }
+
+    // Currently we are allocating three 4K pages for ACPI tables. We are allocating using the
+    // FirstMatch policy, with an 1 byte alignment. This test checks that we are able to allocate
+    // up to this size, and get back the expected addresses.
+    #[test]
+    fn test_write_acpi_table_memory_allocation() {
+        // A mocke Vmm object with 128MBs of memory
+        let mut vmm = default_vmm();
+
+        // This should succeed
+        let mut sdt = MockSdt(vec![0; 4096]);
+        assert!(matches!(vmm.write_acpi_table(&mut sdt), Ok(ACPI_MEM_START)));
+
+        // Let's try to write two 4K pages plus one byte
+        let mut sdt = MockSdt(vec![0; 8193]);
+        assert!(matches!(
+            vmm.write_acpi_table(&mut sdt),
+            Err(AcpiError::VmAllocator(
+                vm_allocator::Error::ResourceNotAvailable
+            ))
+        ));
```

Comment:
nit(applies in a few other places as well, but only commenting here): Just doing `assert!(matches!())` won't print a very nice failure message should the assert fail (it'll just dump some ill-formatted rust syntax and say "assertion failed").

```suggestion
        let err = vmm.write_acpi_table(&mut sdt).unwrap_err();
        assert!(matches!(
            err,
            Err(AcpiError::VmAllocator(
                vm_allocator::Error::ResourceNotAvailable
            ))
        ), "{:?}", err);
```

Reply:
Fixed these wherever I found them. Please let me know if I missed any.